changes/2023-09-26_encryption-context-decrypt/background.md

```diff
@@ -0,0 +1,540 @@
+   frameLength := None
+));
+
+expect encryptOutput.Success?;
+var esdkCiphertext := encryptOutput.value.ciphertext;
+
+defaultCMM :- expect mpl.CreateDefaultCryptographicMaterialsManager(
+   mplTypes.CreateDefaultCryptographicMaterialsManagerInput(
+         keyring := aesKeyring
+   )
+);
+
+// Create Required EC CMM with the required EC Keys we want
+reqCMM :- expect mpl.CreateRequiredEncryptionContextCMM(
+   mplTypes.CreateRequiredEncryptionContextCMMInput(
+         underlyingCMM := Some(defaultCMM),
+         keyring := None,
+         requiredEncryptionContextKeys := requiredEncryptionContextKeys
+   )
+);
+
+var decryptOutput := esdk.Decrypt(Types.DecryptInput(
+   ciphertext := esdkCiphertext,
+   materialsManager := Some(reqCMM),
+   keyring := None,
+   encryptionContext := Some(reproducedEncryptionContext)
+));
+
+expect decryptOutput.Success?;
+var cycledPlaintext := decryptOutput.value.plaintext;
+expect cycledPlaintext == asdf;
+```
+
+### Test Removing Encryption Context on Encrypt is Backwards Compatible
+
+On Encrypt will write all encryption context supplied to the message.
+On Decrypt will have a Required Encryption Context CMM that requires all encryption context
+on Decrypt.
+
+```
+// The string "asdf" as bytes
+var asdf := [ 97, 115, 100, 102 ];
+var aesKeyring := GetAesKeyring();
+
+// Test supply same encryption context on encrypt and decrypt NO filtering
+var encryptionContext := map[ keyA := valA, keyB := valB ];
+var requiredEncryptionContextKeys := [keyA, keyB];
+var reproducedEncryptionContext := map[ keyA := valA, keyB := valB];
+
+var defaultCMM :- expect mpl.CreateDefaultCryptographicMaterialsManager(
+   mplTypes.CreateDefaultCryptographicMaterialsManagerInput(
+         keyring := aesKeyring
+   )
+);
+
+var encryptOutput := esdk.Encrypt(Types.EncryptInput(
+   plaintext := asdf,
+   encryptionContext := Some(encryptionContext),
+   materialsManager := Some(defaultCMM),
+   keyring := None,
+   algorithmSuiteId := None,
+   frameLength := None
+));
+
+expect encryptOutput.Success?;
+var esdkCiphertext := encryptOutput.value.ciphertext;
+
+// Create Required EC CMM with the required EC Keys we want
+var reqCMM :- expect mpl.CreateRequiredEncryptionContextCMM(
+   mplTypes.CreateRequiredEncryptionContextCMMInput(
+         underlyingCMM := Some(defaultCMM),
+         keyring := None,
+         requiredEncryptionContextKeys := requiredEncryptionContextKeys
+   )
+);
+
+var decryptOutput := esdk.Decrypt(Types.DecryptInput(
+   ciphertext := esdkCiphertext,
+   materialsManager := Some(reqCMM),
+   keyring := None,
+   encryptionContext := Some(reproducedEncryptionContext)
+));
+
+expect decryptOutput.Success?;
+var cycledPlaintext := decryptOutput.value.plaintext;
+expect cycledPlaintext == asdf;
+```
+
+### Test Different Encryption Context on Decrypt Failure
+
+Encrypt with and store all encryption context in the message header.
+On Decrypt will supply additional encryption context not stored in the header.
```

Comment:
This is testing a total disjoint between expected EC and inputted EC. We should seperately test "inputted EC is missing a pair", "inputted EC has an incorrect value for one pair", and "inputted EC has an extra pair".

Reply:
Looking at the tests below, I think we cover all of these except "inputted EC has an extra pair"

Reply:
adding in latest revision